nx/source/runtime/diag.c

```diff
@@ -0,0 +1,341 @@
+    if(diagIsChunkTypeEmpty(&chunk_type->header)) {
+        return payload_buf;
+    }
+    u8 *buf = diagLogPayloadEncode(payload_buf, &chunk_type->header, sizeof(DiagLogDataChunkTypeHeader));
+    // Copy value.
+    return diagLogPayloadEncode(buf, &chunk_type->value, chunk_type->header.chunk_len);
+}
+
+NX_CONSTEXPR u8 *diagLogPayloadEncodeStringChunkType(u8 *payload_buf, DiagStringChunkType *chunk_type) {
+    if(diagIsChunkTypeEmpty(&chunk_type->header)) {
+        return payload_buf;
+    }
+    u8 *buf = diagLogPayloadEncode(payload_buf, &chunk_type->header, sizeof(DiagLogDataChunkTypeHeader));
+    // Copy string by its length.
+    return diagLogPayloadEncode(buf, chunk_type->value, chunk_type->header.chunk_len);
+}
+
+void diagLogImpl(const DiagLogMetadata *metadata) {
+    mutexLock(&g_logMutex);
+    Result rc = smInitialize();
+    if(R_SUCCEEDED(rc)) {
+        rc = lmInitialize();
+        smExit();
+    }
+    if(R_SUCCEEDED(rc)) {
+        size_t packet_count = 0;
+        const size_t text_log_len = __builtin_strlen(metadata->text_log);
+        DiagLogPacket *packets = diagAllocateLogPackets(text_log_len, &packet_count);
+        if(packet_count > 0 && packets != NULL) {
+            DiagLogPacket *head_packet = &packets[0];
+            head_packet->header.flags = DiagLogPacketFlags_Head;
+            DiagLogPacket *tail_packet = &packets[packet_count - 1];
+            tail_packet->header.flags = DiagLogPacketFlags_Tail;
+
+            const size_t file_name_len = __builtin_strlen(metadata->source_info.file_name);
+            diagStringChunkTypeSetValue(&head_packet->payload.file_name, DiagLogDataChunkKey_FileName, metadata->source_info.file_name, file_name_len);
+
+            const size_t function_name_len = __builtin_strlen(metadata->source_info.function_name);
+            diagStringChunkTypeSetValue(&head_packet->payload.function_name, DiagLogDataChunkKey_FunctionName, metadata->source_info.function_name, function_name_len);
+
+            diagU32ChunkTypeSetValue(&head_packet->payload.line_number, DiagLogDataChunkKey_LineNumber, metadata->source_info.line_number);
+
+            // Set libnx as module name.
+            const char *module = "libnx";
+            diagStringChunkTypeSetValue(&head_packet->payload.module_name, DiagLogDataChunkKey_ModuleName, module, __builtin_strlen(module));
+
+            size_t remaining_len = text_log_len;
+            DiagLogPacket *cur_packet = head_packet;
+            const char *text_log_buf = metadata->text_log;
+            while(remaining_len > 0) {
+                const size_t cur_len = (remaining_len > DIAG_MAX_STRING_LEN) ? DIAG_MAX_STRING_LEN : remaining_len;
+                diagStringChunkTypeSetValue(&cur_packet->payload.text_log, DiagLogDataChunkKey_TextLog, text_log_buf, cur_len);
+                cur_packet++;
+                text_log_buf += cur_len;
+                remaining_len -= cur_len;
+            }
+
+            size_t i;
```

Comment:
As previously mentioned, move this into the for loop.

Reply:
Oh ok, I thought C didn't allow creating the var inside the loop so I did it that way